src/librustc_trans/back/link.rs

```diff
@@ -168,7 +168,9 @@ pub(crate) fn link_binary(sess: &Session,
     if !sess.opts.cg.save_temps {
         if sess.opts.output_types.should_trans() {
             for obj in trans.modules.iter().filter_map(|m| m.object.as_ref()) {
-                remove(sess, obj);
+                if !preserve_objects_for_their_debuginfo(sess) {
```

Comment:
Not that it makes much of a difference but this `if` could be pulled out of the loop, I think.